Source/Diagnostics/SliceDiagnostic.cpp

```diff
@@ -0,0 +1,490 @@
+ *  \param SliceType is the index type of the slice. By default is slice is cell-centered. 
+ *  If the source multifab (mf) has a different index type then SliceType is converted.
+ */
+
+using namespace amrex;
+std::unique_ptr<MultiFab> CreateSlice( const amrex::MultiFab& mf, 
+                          const amrex::Vector<Geometry> &dom_geom, 
+                          amrex::RealBox &slice_realbox, 
+                          amrex::IntVect &slice_cr_ratio )
+{
+    std::unique_ptr<MultiFab> smf;
+    std::unique_ptr<MultiFab> cs_mf;
+
+    Vector<int> slice_ncells(AMREX_SPACEDIM);
+    int nghost = 1;
+    int nlevels = dom_geom.size();
+    int ncomp = (mf).nComp();
+    
+    IntVect cr_ratio(AMREX_D_DECL(0,0,0));
+    const auto conversionType = (mf).ixType();
+    IntVect SliceType(AMREX_D_DECL(0,0,0));
+ 
+    for (int idim = 0; idim < AMREX_SPACEDIM; ++idim )
+    {
+        if ( conversionType.nodeCentered(idim))
+        {
+           SliceType[idim] = 1;
+        }
+    }
+
+    const RealBox& real_box = dom_geom[0].ProbDomain();
+    RealBox slice_cc_nd_box;
+    int slice_grid_size = 32;
+         
+    bool interpolate = false;
+    bool coarsen = false;
+
+    // same index space as domain //
+    IntVect slice_lo(AMREX_D_DECL(0,0,0));
+    IntVect slice_hi(AMREX_D_DECL(1,1,1));
+    IntVect slice_lo2(AMREX_D_DECL(0,0,0));
+    IntVect interp_lo(AMREX_D_DECL(0,0,0));
+
+    CheckSliceInput(real_box, slice_cc_nd_box, slice_realbox, cr_ratio, 
+                    slice_cr_ratio, dom_geom, SliceType, slice_lo, 
+                    slice_hi, interp_lo);
+
+    // Determine if interpolation is required and number of cells in slice //
+    for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {    
+       // Flag for interpolation if required //
+       if ( interp_lo[idim] == 1) {
+          interpolate = 1;
+       }
+
+       // For the case when a dimension is reduced //
+       if ( ( slice_hi[idim] - slice_lo[idim]) == 1) {       
+          slice_ncells[idim] = 1;
+       }
+       else {       
+          slice_ncells[idim] = ( slice_hi[idim] - slice_lo[idim] + 1 ) 
+                                / cr_ratio[idim];
+
+          int refined_ncells = (slice_hi[idim] - slice_lo[idim]) + 1 ;
+          if ( cr_ratio[idim] > 1) {          
+             coarsen = true;
+
+             // modify slice_grid_size if >= refines_cells //
+             if ( slice_grid_size >= refined_ncells ) {
+                slice_grid_size = refined_ncells - 1;
+             }
+
+          }
+       }
+    }
+
+    // Slice generation with index type inheritance //
+    Box slice(slice_lo, slice_hi);
+   
+    // Convert from cc to index type of parent multifab //
+    bool slicetypeToBeConverted = 0;
+    for ( int idim = 0; idim < AMREX_SPACEDIM; ++idim ) {    
+       if ( SliceType[idim] == 1 ) {
+          slicetypeToBeConverted = 1;
+       }
+    }
+
+    Vector<BoxArray> sba(1);
```

Comment:
I declared it as Vector<> since in the future we may need to use this for a multi-level AMR grid. 
So I tried to keep the arrays in place for it -- although this PR assumes a single level grid.

Reply:
That makes sense, thank you